src/org/sosy_lab/java_smt/basicimpl/AbstractQuantifiedFormulaManager.java

```diff
@@ -66,124 +56,163 @@ private BooleanFormula wrap(TFormulaInfo formulaInfo) {
   @Override
   public BooleanFormula eliminateQuantifiers(BooleanFormula pF)
       throws InterruptedException, SolverException {
-    List<ProverOptions> proverOptions = extractQuantifierEliminationOptions();
-    if (proverOptions.contains(ProverOptions.SOLVER_INDEPENDENT_QUANTIFIER_ELIMINATION)) {
-      try {
-        return wrap(eliminateQuantifiersUltimateEliminator(pF));
-      } catch (UnsupportedOperationException | IllegalArgumentException e) {
-        if (proverOptions.contains(ProverOptions.QUANTIFIER_ELIMINATION_FALLBACK_WARN_ON_FAILURE)) {
+    return wrap(eliminateQuantifiers(extractInfo(pF)));
+  }
+
+  @Override
+  public BooleanFormula eliminateQuantifiers(BooleanFormula pF, QuantifierEliminationMethod pMethod)
+      throws InterruptedException, SolverException {
+    assert pMethod != null : "Quantifier elimination method cannot be null";
+    switch (pMethod) {
+      case ULTIMATE_ELIMINATOR:
+        try {
+          return wrap(eliminateQuantifiersUltimateEliminator(pF));
+        } catch (UnsupportedOperationException | IllegalArgumentException e) {
+          logger.logException(
+              Level.SEVERE,
+              e,
+              "UltimateEliminator failed. Please adjust the "
+                  + "option if you want to use the native quantifier elimination");
+
+          throw e;
+        }
+
+      case ULTIMATE_ELIMINATOR_FALLBACK_ON_FAILURE:
+        try {
```

Comment:
You are repeating the try-catch blocks quite a lot. You could extract them into a method.